crates/abq_utils/src/net_protocol/runners.rs

```diff
@@ -289,38 +289,60 @@ impl From<RawManifest> for Manifest {
 impl Manifest {
     /// Flattens a manifest into [TestSpec]s, preserving the manifest order.
     pub fn flatten(self) -> (Vec<TestSpec>, MetadataMap) {
-        use v0_2::{Group, Test, TestCase, TestOrGroup};
-
+        use v0_2::{Group, TestOrGroup};
         let Manifest { members, init_meta } = self;
         let mut collected = Vec::with_capacity(members.len());
-        let mut queue: VecDeque<_> = members.into_iter().collect();
-        while let Some(test_or_group) = queue.pop_front() {
-            match test_or_group {
-                TestOrGroup::Test(Test { id, meta, .. }) => {
-                    let spec = TestSpec {
-                        // Generate a fresh ID for ABQ-internal usage.
-                        work_id: WorkId::new(),
-                        test_case: self::TestCase(
-                            TestCase {
-                                id,
-                                meta,
-                                focus: None,
-                            }
-                            .into(),
-                        ),
-                    };
-                    collected.push(spec);
+
+        // share a top-level vector so we can reuse the largest queue allocated
+        let mut queue: Vec<v0_2::TestOrGroup> = Vec::new();
+        for top_level_test_or_group in members {
+            let group_id = GroupId::new();
+            match top_level_test_or_group {
+                TestOrGroup::Test(test) => {
+                    Self::add_test_to_collected(test, &mut collected, group_id)
                 }
                 TestOrGroup::Group(Group { members, .. }) => {
-                    for member in members.into_iter().rev() {
-                        queue.push_front(member);
+                    queue.clear();
+                    queue.extend(members);
```

Comment:
I think we need to extend the queue with the members in reverse. The reason is that otherwise we will visit the first member last.